### PR TITLE
Add janus model

### PR DESCRIPTION
**Pull request by hsilva664**
Small bug fixes

### CODE REVIEW

File: src/transformers/models/janus/convert_janus_weights_to_hf.py [hunk 1]
```diff
@@ -102,6 +102,9 @@ def convert_key(old_key, vision_mappings, other_mappings):
 
 
 def convert_state_dict(old_state_dict, vision_mappings, other_mappings):
+    """TODO
+    Has to delete the parameter `vqmodel.quantize.codebook_used`.
```
Comment (hsilva664) [inline src/transformers/models/janus/convert_janus_weights_to_hf.py:106]: This was causing me some errors when importing. I think this parameter has to do with the percentage of the codebook that was actually used, which they report in the paper, but it does not seem to be used in the code.

File: src/transformers/models/janus/modular_janus.py [hunk 2]
```diff
@@ -1022,4 +1025,7 @@ def resize(
         image = to_numpy_array(image)
         return image
 
-__all__ = ["JanusImageProcessor",]
+__all__ = ["JanusImageProcessor",
```
Comment (hsilva664) [inline src/transformers/models/janus/modular_janus.py:1028]: This was causing me importing errors when missing

File: tmp/original_implementation/test_logits.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+"""
```
Comment (hsilva664) [inline tmp/original_implementation/test_logits.py:1]: This is just the old code that I was using for my previous VQ implementation, may delete later

File: tmp/new_implementation/test_logits.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
+"""
```
Comment (hsilva664) [inline tmp/new_implementation/test_logits.py:1]: This is the new code used to reproduce logits

File: tmp/new_implementation/test_logits.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+import torch
+import numpy as np
+from transformers import enable_full_determinism, JanusForConditionalGeneration, JanusConfig
+
+
+def main(args):
+    enable_full_determinism(0)
+
+    # Emu3 example also loads this class like this in tests and in the docs, instead of using Auto
+    model = JanusForConditionalGeneration(JanusConfig(vq_config={"resolution": 384}))
```
Comment (hsilva664) [inline tmp/new_implementation/test_logits.py:21]: This was causing a shape mismatch problem. The config default parameters assumed a default resolution of 512. However, the example code generates only 576 tokens, which correspond to a raw resolution of 384 (prior to the downsizing resulting from convolutions).

File: src/transformers/models/janus/modular_janus.py [hunk 6]
```diff
@@ -721,12 +724,12 @@ def forward(self, hidden_state: torch.FloatTensor) -> torch.FloatTensor:
         hidden_state = self.mid.block_2(hidden_state)
 
         # upsampling
-        for i_level in reversed(range(self.num_resolutions)):
```
Comment (hsilva664) [inline src/transformers/models/janus/modular_janus.py:724]: When creating this outermost module list in the __init__, you created it iterating in reverse. However, if you look at original implementation, even though they iterate in regular order, your hidden sizes occur in the same positions as theirs (e.g. 512 occurs in position 0 in both codebases). Therefore, this loop has to be iterated in normal order. This solves a shape mismatch bug.

File: src/transformers/models/janus/modeling_janus.py [hunk 7]
```diff
@@ -953,6 +953,7 @@ def get_codebook_entry(self, image_tokens: torch.LongTensor) -> torch.FloatTenso
         emb_dim: int = self.embedding.weight.shape[-1]
         # get quantized latent vectors
         hidden_state_quant = self.embedding(image_tokens)
+        hidden_state_quant = F.normalize(hidden_state_quant, p=2, dim=-1)
```
Comment (hsilva664) [inline src/transformers/models/janus/modeling_janus.py:956]: On the original codebase, they keep this in an IF-ELSE statement with a parameter to control whether normalization happens or not. However, it is set to true, so the normalization should occur for logits to match.

File: src/transformers/models/janus/modular_janus.py [hunk 8]
```diff
@@ -559,6 +559,9 @@ def get_codebook_entry(self, image_tokens: torch.LongTensor) -> torch.FloatTenso
         emb_dim: int = self.embedding.weight.shape[-1]
         # get quantized latent vectors
         hidden_state_quant = self.embedding(image_tokens)
+        # l2 normalization on the last dimension
+        hidden_state_quant = F.normalize(hidden_state_quant, p=2, dim=-1)
```
Comment (hsilva664) [inline src/transformers/models/janus/modular_janus.py:563]: On the original codebase, they keep this in an IF-ELSE statement with a parameter to control whether normalization happens or not. However, it is set to true, so the normalization should occur for logits to match.

File: tmp/new_implementation/test_logits.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    # assert torch.isclose(dec.cpu(), answers["dec"], atol=1e-3).all()
+
+    codebook_entry = model.vqmodel.quantize.get_codebook_entry(generated_tokens.to(dtype=torch.int))
+    assert torch.isclose(codebook_entry.cpu(), answers["dec_quant_b"], atol=1e-3).all()
+    hidden_states = model.vqmodel.post_quant_conv(codebook_entry)
+    assert torch.isclose(hidden_states.cpu(), answers["dec_quant"], atol=1e-3).all()
+    pixel_values = model.vqmodel.decoder(answers["dec_quant"].cuda())
+    # This does not match. I believe it is due to different cuda versions between my docker container to run Janus
+    # and the env in which I was developing transformers
+    assert torch.isclose(pixel_values.cpu(), answers["dec_dec"], atol=1e-3).all()
```
Comment (hsilva664) [inline tmp/new_implementation/test_logits.py:91]: This last part did not match successfully on my test yet, but I think this might be due to different CUDA versions, see if it matches for you.

**Comment by yaswanth19** [comment]
@hsilva664 Use this latest chat template. With this we are matching the SFT template of Janus.

{
    "chat_template": "{% set seps = ['\\n\\n', '<｜end▁of▁sentence｜>'] %}{% for message in messages %}{{ message['role'].capitalize() + ': ' + message['content'] + seps[loop.index0 % 2] if message['content'] else message['role'].capitalize() + ':' }}{% endfor %}"
}

messages = [
    {
        "role": "User",
        "content": "<image_placeholder>\nConvert the formula into latex code.",
    },
    {"role": "Assistant", "content": ""},
]

Here we expect the user to provide image_placeholder so no content type == image/text is necessary.
